mobile/src/main/java/org/openhab/habdroid/ui/ChartWidgetActivity.kt

```diff
@@ -104,104 +146,368 @@ class ChartWidgetActivity :
                 onRefresh()
                 true
             }
-            item.itemId == R.id.show_legend -> {
-                showLegend = !showLegend
-                updateHasLegendButtonState(item)
-                true
-            }
             item.itemId == android.R.id.home -> {
                 finish()
                 super.onOptionsItemSelected(item)
             }
             // The dropdown menu is opened
-            item.itemId == R.id.period -> true
-            periodForMenuItem(item.itemId) != null -> {
-                period = periodForMenuItem(item.itemId)!!
+            item.itemId == R.id.period -> {
+                true
+            }
+            DURATION_MENU_MAPPING.containsKey(item.itemId) -> {
+                period = DURATION_MENU_MAPPING[item.itemId]!!
                 onRefresh()
                 true
             }
             else -> super.onOptionsItemSelected(item)
         }
     }
 
-    private fun periodForMenuItem(itemId: Int) = when (itemId) {
-        R.id.period_h -> "h"
-        R.id.period_4h -> "4h"
-        R.id.period_8h -> "8h"
-        R.id.period_12h -> "12h"
-        R.id.period_d -> "D"
-        R.id.period_2d -> "2D"
-        R.id.period_3d -> "3D"
-        R.id.period_w -> "W"
-        R.id.period_2w -> "2W"
-        R.id.period_m -> "M"
-        R.id.period_2m -> "2M"
-        R.id.period_4m -> "4M"
-        R.id.period_y -> "Y"
-        else -> null
-    }
-
     override fun onSaveInstanceState(outState: Bundle) {
-        outState.putBoolean(SHOW_LEGEND, showLegend)
-        outState.putString(PERIOD, period)
+        when (val p = period) {
+            is Period -> outState.putSerializable(PERIOD, p)
+            is Duration -> outState.putSerializable(PERIOD, p)
+        }
+        outState.putParcelable(DATA, loadedChartData)
         super.onSaveInstanceState(outState)
     }
 
-    override fun onRestoreInstanceState(savedInstanceState: Bundle) {
-        showLegend = savedInstanceState.getBoolean(SHOW_LEGEND)
-        period = savedInstanceState.getString(PERIOD)!!
-        super.onRestoreInstanceState(savedInstanceState)
+    override fun onStart() {
+        super.onStart()
+        val data = loadedChartData
+        val dataIsOutdated = data != null &&
+            widget.refresh > 0 &&
+            Duration.between(data.timestamp, Instant.now()).toMillis() > widget.refresh
+        if (data == null || dataIsOutdated) {
```

Comment:
I wonder if I should consider data usage policy here?

Reply:
Ok, than I'd avoid refreshing when data saver is enabled.